Rank recipients: support min/max number of choices #8009

VamsiSangam opened the pull request:
Fixes #8009 

- Added min/max recipients to be ranked for Rank recipient question.
- Slightly modified min/max options for Rank options feature (eg. constant names, variable names, etc.)

Status - ongoing.

VamsiSangam commented:
UI changes -

![1](https://user-images.githubusercontent.com/11662001/30774133-5ac2b760-a09b-11e7-9757-f5c39cebd503.JPG)
![2](https://user-images.githubusercontent.com/11662001/30774136-5dae1abe-a09b-11e7-9f56-dfea66ddd6a9.JPG)

![3](https://user-images.githubusercontent.com/11662001/30774141-62a7b1e2-a09b-11e7-9053-eccf49f26412.JPG)

![4](https://user-images.githubusercontent.com/11662001/30774144-6a12ba44-a09b-11e7-8ce7-b136b10f1dcf.JPG)

I'm not very confident about the look of the UI :sweat_smile: . @damithc sir, could you please give your inputs regarding the look and feel of the new UI?

Ready for review.

damithc commented:
Do we need to repeat this for each option?
![image](https://user-images.githubusercontent.com/1673303/30774324-5358547c-a0b3-11e7-9055-d57de0b1ebd7.png)

VamsiSangam commented:
The current system implementation doesn't have a way of avoiding it sir. I was planning to add new methods which will enable any question to display some additional information here -

![4](https://user-images.githubusercontent.com/11662001/30774724-e8b3a346-a0a4-11e7-82d5-4ee5c3bdcd17.JPG)


That is, just before the recipients and just after visibility hint. With this, we can mention the min/max recipients to be ranked there. Shall I proceed with this approach sir?

damithc commented:
> That is, just before the recipients and just after visibility hint. With this, we can mention the min/max recipients to be ranked there. Shall I proceed with this approach sir?

Sure, let's give that a try.

VamsiSangam commented:
![capture](https://user-images.githubusercontent.com/11662001/30813057-fb0567ec-a229-11e7-9f51-874159e7546b.JPG)

@damithc sir, is this UI acceptable? The code is written such that any question can choose (if needed) to display instructions. Also, is the instructor feedback edit page UI acceptable?

VamsiSangam commented:
Ready for review.

wkurniawan07 commented on src/main/webapp/dev/js/common/questionRank.es6:
```diff
@@ -2,24 +2,55 @@ import {
+    } else if (recipient === 'OWN_TEAM_MEMBERS_INCLUDING_SELF') {
+        // returning infinite as this is dependent on team size
+        return Number.MAX_SAFE_INTEGER;
+    }
```
You can shorten this block:
```js
switch (recipient) {
case 'STUDENTS':
case 'INSTRUCTORS':
case 'TEAMS':
    return $(`#num-${recipient.toLowerCase()}`).val();
case 'OWN_TEAM_MEMBERS':
case 'OWN_TEAM_MEMBERS_INCLUDING_SELF':
    return Number.MAX_SAFE_INTEGER;
default:
    return 1;
}
```

wkurniawan07 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -888,10 +888,19 @@ function updateRankMessageQn(qnNum) {
+                                  return $(this).val() !== '';
+                              }).length;
+        } else {
+            rankedOptions = $(`select[name="responsetext-${questionNumber}-${recipientIndex}"]`)
```
If the difference is only in the selector name, consider confining the `if` condition to just that.
```js
const selector = $(`#rankToRecipients-${questionNumber}`).val() === 'true' ? ... : ...;
const rankedOptions = $(...).filter(...).length;
```

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java:
```diff
@@ -1554,24 +1554,58 @@ public void clickRemoveRankOptionLink(int qnIndex, int optionIndex) {
+            WebElement rankOptionsTable = browser.driver.findElement(By.id("rankOptionTable-" + qnIndex));
+            List<WebElement> optionInputFields = rankOptionsTable
+                                                    .findElements(
+                                                         By.cssSelector("input[id^='rankOption-']"));
```
Not your fault, but let's out everything after `=` on the next line.
```java
List<WebElement> optionInputFields =
        rankOptionsTable...
```

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java:
```diff
@@ -1554,24 +1554,58 @@ public void clickRemoveRankOptionLink(int qnIndex, int optionIndex) {
+            return Integer.MAX_VALUE;
+        default:
+            // For SELF, OWN_TEAM, NONE
+            return 1;
```
Here you see how my previous suggestion works. You can make it even more concise by allowing fall-throughs.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackQuestionDetails.java:
```diff
@@ -111,6 +111,14 @@ public String getQuestionText() {
 
     /**
+     * Returns a list of strings where each string is an instruction to answer the question.
+     * This will be displayed in the submission page just above the responses.
```
This should not be there. The method has no business as to where the instructions will be displayed.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRankRecipientsQuestionDetails.java:
```diff
@@ -74,15 +74,15 @@ public String getQuestionWithExistingResponseSubmissionFormHtml(
                 Slots.RANK_ARE_DUPLICATES_ALLOWED_VALUE, Boolean.toString(isAreDuplicatesAllowed()),
                 Slots.RANK_IS_MAX_OPTIONS_TO_BE_RANKED_ENABLED, isMaxOptionsToBeRankedEnabled ? "" : "disabled",
-                Slots.RANK_DISPLAY_MAX_OPTIONS_HINT, isMaxOptionsToBeRankedEnabled ? "" : "hidden",
+                Slots.RANK_DISPLAY_MAX_OPTIONS_HINT, "hidden",
```
If this is no longer a variable, why not hard-code `Slots.RANK_DISPLAY_MAX_OPTIONS_HINT` in the template itself?

VamsiSangam replied:
Rank recipients and Rank options share the same submission page template, for Rank options we show the hint, and for Rank recipients, we hide it and show instructions instead. Also, the flexibility of using `hidden` to hide the hints is useful in cases where both min/max are disabled or either of them is disabled, in which case we hide whatever is disabled. So, `hidden` is pretty useful here commonly used by Rank options and recipients.

wkurniawan07 commented on src/main/resources/feedbackQuestionRankRecipientsEditFormTemplate.html:
```diff
@@ -8,6 +8,42 @@
+      <div class="col-sm-6 col-xs-12">
+        <div class="pull-left">
+          <input type="number" class="form-control" name="${Const.ParamsNames.FEEDBACK_QUESTION_RANKMINRECIPIENTSTOBERANKED}"
+            id="${Const.ParamsNames.FEEDBACK_QUESTION_RANKMINRECIPIENTSTOBERANKED}-${questionNumber}" min="1" value="${minRecipientsToBeRanked}"/>
```
Need 2 more spaces to the right, and don't close the tag with `/>`.

wkurniawan07 commented on src/main/resources/feedbackQuestionRankRecipientsEditFormTemplate.html:
```diff
@@ -8,6 +8,42 @@
+      <div class="col-sm-6 col-xs-12">
+        <div class="pull-left">
+          <input type="number" class="form-control" name="${Const.ParamsNames.FEEDBACK_QUESTION_RANKMAXRECIPIENTSTOBERANKED}"
+              id="${Const.ParamsNames.FEEDBACK_QUESTION_RANKMAXRECIPIENTSTOBERANKED}-${questionNumber}" min="1" value="${maxRecipientsToBeRanked}"/>
```
Similarly don't close the tag with `/>`.

wkurniawan07 commented on src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java:
```diff
@@ -1554,24 +1554,58 @@ public void clickRemoveRankOptionLink(int qnIndex, int optionIndex) {
 
     private WebElement getMinOptionsToBeRankedCheckbox(int qnNumber) {
-        return browser.driver.findElement(By.id("minOptionsToBeRankedEnabled-" + qnNumber));
+        if ("RANK_OPTIONS".equals(getQuestionType(qnNumber))) {
```
`"RANK_OPTIONS".equals(getQuestionType(qnNumber))` should be extracted out.

wkurniawan07 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackRankRecipientsQuestionDetails.java:
```diff
@@ -157,13 +172,30 @@ private String getSubmissionOptionsHtmlForRankingRecipients(int totalNumRecipien
+                Slots.RANK_PARAM_MIN_RECIPIENTS_TO_BE_RANKED,
+                        Const.ParamsNames.FEEDBACK_QUESTION_RANKMINRECIPIENTSTOBERANKED,
+                Slots.RANK_MIN_RECIPIENTS_TO_BE_RANKED, isMinOptionsToBeRankedEnabled
+                        ? Integer.toString(minOptionsToBeRanked) : "1",
```
May look more readable if `isMinOptionsToBeRankedEnabled` is also on the next line. Similarly for the `isMaxOptionsToBeRankedEnabled` down there.

wkurniawan07 replied:
@VamsiSangam take a look at the comment before this

wkurniawan07 replied:
I mentioned "if `isMinOptionsToBeRankedEnabled` is also on the next line", which is:
```java
Slots.RANK_MIN_RECIPIENTS_TO_BE_RANKED,
        isMinOptionsToBeRankedEnabled ? Integer.toString(minOptionsToBeRanked) : "1",
```
Then again it's just cosmetic, feel free to just ignore it.

wkurniawan07 replied:
@VamsiSangam you're targeting the wrong line... you fixed the one that doesn't need fix. It's the one above this (let me just be precise: line 207, and also line 215 down there).

wkurniawan07 commented on src/main/java/teammates/common/util/Templates.java:
```diff
@@ -411,21 +411,30 @@ public static String populateTemplate(String template, String... keyValuePairs)
+            public static final String RANK_DISPLAY_MIN_RECIPIENTS_HINT = "${displayMinRecipientsHint}";
+            public static final String RANK_DISPLAY_MAX_RECIPIENTS_HINT = "${displayMaxRecipientsHint}";
+            public static final String RANK_IS_MIN_RECIPIENTS_TO_BE_RANKED_ENABLED = "${isMinRecipientsToBeRankedEnabled}";
+            public static final String RANK_IS_MAX_RECIPIENTS_TO_BE_RANKED_ENABLED = "${isMaxRecipientsToBeRankedEnabled}";
```
Let's make the order uniform: `min-*s`, `max-*s`, `is-min-*s`, `is-max-*s`, `display-min-*s`, `display-max-*s`. You can tweak the order, but the idea is to make it similar in both options and recipients.

wkurniawan07 commented on src/main/java/teammates/common/util/Templates.java:
```diff
@@ -411,21 +411,30 @@ public static String populateTemplate(String template, String... keyValuePairs)
                     "${Const.ParamsNames.FEEDBACK_QUESTION_RANKISDUPLICATESALLOWED}";
+            public static final String RANK_PARAM_OPTION = "${Const.ParamsNames.FEEDBACK_QUESTION_RANKOPTION}";
+            public static final String RANK_PARAM_NUMBER_OF_CHOICE_CREATED =
+                    "${Const.ParamsNames.FEEDBACK_QUESTION_NUMBEROFCHOICECREATED}";
```
May not be related, but are these two params still used?

VamsiSangam replied:
Yes these seem to be used. [FEEDBACK_QUESTION_RANKOPTION](https://github.com/TEAMMATES/teammates/search?utf8=%E2%9C%93&q=FEEDBACK_QUESTION_RANKOPTION&type=) [FEEDBACK_QUESTION_NUMBEROFCHOICECREATED](https://github.com/TEAMMATES/teammates/search?utf8=%E2%9C%93&q=FEEDBACK_QUESTION_NUMBEROFCHOICECREATED&type=)

wkurniawan07 commented on src/test/resources/pages/instructorFeedbackSubmitPageRankHelper.html:
```diff
@@ -121,10 +121,10 @@
               <div>
                 <p class="text-muted" hidden="">
-                  You need to rank at least  options.
+                  You need to rank at least  ${optionsOrRecipients}.
```
Not a big deal from user's POV since this is hidden, but from a tester's POV `${...}` is most of the time associated with God mode. See if you can still inject some appropriate values here.

VamsiSangam replied:
I had a similar feeling, will look into it.

VamsiSangam commented:
Ready for review.

wkurniawan07 commented on src/main/webapp/dev/js/common/questionRank.es6:
```diff
@@ -169,25 +197,29 @@ function removeRankOption(index, questionNum) {
+            const questionNum = $(e.target).closest('form').attr('data-qnnumber');
+            toggleMinOptionsToBeRanked(questionNum);
+        },
+    );
```
This doesn't look right...
```js
$(document).on('change',
        'input...', (e) => {
   ...;
   ...;
});
```
Similarly down there.

wkurniawan07 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -888,10 +888,10 @@ function updateRankMessageQn(qnNum) {
-                              .filter(function () {
-                                  return $(this).val() !== '';
-                              }).length;
+        const selector = $(`#rankToRecipients-${questionNumber}`).val() === 'true' ?
```
The `?` should be on the next line, as for the case of `:` afterwards.

VamsiSangam commented:
Addressing your review -

> Why not for both cases just show instructions?

In the case of rank options, displaying hint for each recipient kinda looks appropriate, at least that's what I feel. But the same is not for rank recipients, so I put instructions for it. @damithc sir do you feel it is better to display instructions for Rank options also instead of displaying hint for each recipient like this -

![capture](https://user-images.githubusercontent.com/11662001/31009218-5d30ea48-a524-11e7-878c-59afc38ac69e.JPG)

> Don't forget this.

I couldn't understand whats wrong. Can you elaborate?

damithc commented:
> In the case of rank options, displaying hint for each recipient kinda looks appropriate, at least that's what I feel. But the same is not for rank recipients, so I put instructions for it. @damithc sir do you feel it is better to display instructions for Rank options also instead of displaying hint for each recipient like this -
![capture](https://user-images.githubusercontent.com/11662001/31009218-5d30ea48-a524-11e7-878c-59afc38ac69e.JPG)

In this screenshot, you are ranking options _for each evaluee_. That's why it is ok to show the instruction for each evaluee. If the question is simply ranking options (i.e. there is no evaluee e.g. `rank three possible product names for a new product`), then you should be showing instructions once only, not once for each product name.

VamsiSangam commented:
Understood sir. I've decided to remove the hint for each recipient and use only instructions as the same hint would be repeated anyways. @wkurniawan07 can I make this removal of hints in this PR?

wkurniawan07 commented:
@VamsiSangam sure, that is what I suggested anyway.

VamsiSangam commented:
Here are screenshots of the new UI for rank options -

![1](https://user-images.githubusercontent.com/11662001/31017300-a7b51ef6-a545-11e7-9c7b-78ab295b4954.JPG)

There's a subtle issue here for rank recipient question. When the recipient is `Team members` then the recipients are dependent on the size of the team, so in this case, the instructions may get displayed incorrectly.

![2](https://user-images.githubusercontent.com/11662001/31017386-e706238e-a545-11e7-954b-aabcf40b01e2.JPG)

@wkurniawan07 @damithc sir, your thoughts on this?

damithc commented:
> There's a subtle issue here for rank recipient question. When the recipient is Team members then the recipients are dependent on the size of the team, so in this case, the instructions may get displayed incorrectly.

We can improve instructions as a separate PR (e.g. give additional instructions if the team size is relevant, not show instructions in dimmed text etc.). We need to finish this PR quickly because it is holding up the next release.

VamsiSangam commented:
Ready for review.

wkurniawan07 commented on src/main/java/teammates/common/util/Templates.java:
```diff
@@ -411,21 +411,28 @@ public static String populateTemplate(String template, String... keyValuePairs)
+            public static final String RANK_IS_MIN_RECIPIENTS_TO_BE_RANKED_ENABLED = "${isMinRecipientsToBeRankedEnabled}";
+            public static final String RANK_IS_MAX_RECIPIENTS_TO_BE_RANKED_ENABLED = "${isMaxRecipientsToBeRankedEnabled}";
+            public static final String RANK_DISPLAY_MIN_RECIPIENTS_HINT = "${displayMinRecipientsHint}";
+            public static final String RANK_DISPLAY_MAX_RECIPIENTS_HINT = "${displayMaxRecipientsHint}";
```
Are the above two constants still used?

VamsiSangam replied:
No, they can be removed :+1:

wkurniawan07 commented on src/main/webapp/dev/js/main/feedbackSubmissionsEdit.es6:
```diff
@@ -888,10 +888,10 @@ function updateRankMessageQn(qnNum) {
-                                  return $(this).val() !== '';
-                              }).length;
+        const selector = $(`#rankToRecipients-${questionNumber}`).val() === 'true'
+                ? `select[name^="responsetext-${questionNumber}-"]`
```
Looks like I missed this. What's the effect of selector with `^=` here?

VamsiSangam replied:
For a rank options question, each recipient has a set of dropdowns and the min/max restrictions impose on that set of dropdowns -

![capture](https://user-images.githubusercontent.com/11662001/31064122-2518f166-a757-11e7-8d1e-68d35f0311db.JPG)

But for rank recipients question, each recipient has just one drop-down, we impose our min/max restrictions over the set of all dropdowns for that question -

![capture](https://user-images.githubusercontent.com/11662001/31064242-1eeaf202-a758-11e7-84ed-623b828f526c.JPG)

so for selecting all dropdowns for the question we use `^=`. The different selectors are for selecting the required set of dropdowns.

VamsiSangam commented:
Ready for review.

VamsiSangam commented:
Ready for review.

wkurniawan07 commented:
@VamsiSangam please revert your last commit d3c03ba; it is not related to this PR, and will be addressed in #8060.

Other than that, the changes are good. @damithc this can already be reviewed.

VamsiSangam commented:
Reverted the last commit.
